### ISSUE TITLE
Unable to manage account

### ISSUE THREAD

**Issue by Ramius1701** [djphil/osregister#5]
Right after login, you get the message that you logged in, click on manage your account and it logs you right back out

**Comment by djphil** [comment]
This is probably caused by an invalid session, but I cannot reproduce the problem.

Try this, goto config.php line 2 and
replaces
// session_start ();
by
session_start ();

hopefully ...

**Comment by Ramius1701** [comment]
Tried both ways and same result

Running:
Apache 2.4.1
PHP 7.4.4
MariaDB 10.4.12

**Comment by djphil** [comment]
I installed osregister from scratch today and still haven't managed to reproduce your problem.
For me all goes well, without error ...

Here's some informations ...
In the account.php page there are 2 places where there is a redirect in case of session error.
1st is line 2
https://github.com/djphil/osregister/blob/master/inc/account.php#L2
The second is from line 57 to 68
https://github.com/djphil/osregister/blob/master/inc/account.php#L57

You could possibly delete these parts one by one to determine which of the 2 redirects you.
This is not a solution but perhaps would allow progress to be made towards a solution ...

You could also check your log files and provide useful information if they contain any.
Unfortunately, I do not see how I could help you further.

**Comment by Ramius1701** [comment]
I was able to get to the account page after commenting out lines 57-68 in the account.php file

**Comment by Ramius1701** [comment]
After some trial and error, i have narrowed it down to commenting out lines 66 and 67

**Comment by djphil** [comment]
Theoretically, this would mean that one of these session identifiers is empty:
username, email, password, created, useruuid, usertitle, active
You could check in your "useraccounts" database if all these fields are correctly filled.
If one of these fields is empty in your database, this could be the source of the problem.

**Comment by Ramius1701** [comment]
Checked the DB and only the column of UserTitle was empty

And I comment out this line:

Line 64:    empty($_SESSION['usertitle']) ||

It works! Now by default, any account created via the opensim.exe console will not input some of this information.

These two lines were already commented out:

Line 62:    // empty($_SESSION['userlevel']) ||
Line 63:    // empty($_SESSION['userflags']) ||

I changed back to default and I created an account with the registration form, all the data is there. I am able to log in with the new account with Firestorm. Try and login with the site and still cannot get to the account page. Checked the database tables and the active column is set to 0. Now I manually edit the column and change it to 1 for that user and now I can access the account.php page.

Creating an account this way doesn't work in creating the balances account when using 

https://github.com/BigManzai/OpenSimCurrencyServer-2020

That's another issue :)

**Comment by djphil** [comment]
Ok your disconnection problem is now solved, it's a good thing.

And yes, by default the accounts created with osregister are not active, they have the value 0. If you want the accounts to be activated by default, you must change the 0 to 1 in the file functions.php on line 284.
https://github.com/djphil/osregister/blob/master/inc/functions.php#L284

For your problem with OpenSimCurrencyServer, I don't know. I do not use. 
If desired, make another detailed report.

**Comment by emperorstarfinder** [comment]
So I thought I would do a quick check on this as we are using osregister temporarily while we are working on our new main website for our grid.  Unfortunately, I cannot reproduce the problems described here either.  This would suggest that this is probably something local to the server environment or configurations are done on the local server.  I tested this both against PHP 7.3.x and 7.4.19 and everything is working fine.

I would probably suggest that the default UserTitle be Resident with documentation on how folks using the code can change it if they choose something more custom related to their grid, and having accounts be active by default.  Also, the version number of the osregister code should not really be included in a grid's name or referenced in relation to a grid in the code and should only be internal as to the code meaning not something that is necessarily visible on the website except in the footer where the creator of the osregister code should be.

As to the OpenSimCurrencyServer according to the README.md file in that repository BigManzai has indicated that using MySqli is not yet working on the website portion of the code so it is probably still not production level ready.  That would be something to use at your own risk.

Hope this helps some.

**Comment by djphil** [comment]
Hello and thank you for your interest for osregister.

I agree with your idea of a local problem probably caused by an improper server configuration. (eg: register global on/off) but I don't know exactly at this point and the problem has not been reported by other osregister users.

Currently the user's title is "Resident" and his account is deactivated by default (see line 274 -> 184 in inc/functions.php) and a small summary documentation has been added in the docs folder. More documentation is always welcome, I grant you that.

In fact, for the activation of accounts, I have planned for a future version of osregister to rework the backend user part and add an administrator backend which will include the list of registered users and allow it to be edited at will. This way it will be easier for the administrator to manage his users.

Compared to the version number I did not understand, can you explain more?

Compared to OpenSimCurrencyServer, I don't think this is an osregister issue so I haven't investigated the issue in this statde.

Having said that, I am happy to hear that osregister is helping you out right now.

**Comment by emperorstarfinder** [comment]
Hi Phil.

I mentioned the bit about the OpenSimCurrencyServer as it was mentioned by the original issue reporter, so I was merely providing an update on this.

As to the version number suggestion, in inc/Config.php.Example. you have on line 28 "$version = "0.1";".  This ends up being included in the website title in the browser and throughout the website itself.  This should really not be doing such and should be more of an internal thing to check if the website code is up to date.  If you don't plan on doing update checks or provide the ability in the admin panel your working on to check for updates to the code, I would suggest removing the checks for the version number throughout the code so it will not appear next to the Website Title information. "i.e. OpenSim Register v1.0"  In most cases for a website you really do to need the version number of the website code being public information but you can have that information provided to the website owner in the admin area.  Hope that clarifies that.

As an additional note, the database documentation on the OpenSim wiki that I had mentioned in a previous issue (was deleted via the last update of GitHub) is in fact correct and does provide the up to date database table information for OpenSim 0.9.x except for a couple of columns that were just added within the last month due to the OpenSim version of E.E.P. (Enhanced Experience Protocol).

Best Regards.

**Comment by Ramius1701** [comment]
Steps to reproduce the problem 

I did not include the MoneyServer with this setup

Brand new install of OpenSim configured for robust and no HG, no regions
New MySQL DB
No changes to the osregister files with the exception of inputting the DB information for the config file

Start Robust

Start OpenSim

At some point you have to create an account to configure your first region, As this is done in the robust console, I used Test User and fill in all the requested information to create the account. Then I inputted all the information to create the first region which I called Welcome

Everything is booted, all is good

Can log in as Test User with Firestorm.

Now you can go to the osregister page and register a new account and log in and everything works!

Can log in as Tester User with Firestorm.

But wait, what if I wanted to log into that first account Test User using osregister, you can log in, but once you click on Manage my account from the main screen or my account from the drop-down you get the message "You are disconnected successfully Test User ..."

This does not happen with Tester User.

Here is why, when you look into the MySQL DB table there is an empty column under useraccounts for the UserTitle for Test User which is required when using the osregister form to login. This does not happen when you create an account with osregister. This comes from creating the user from the robust console. Which after going through the paces set forth in previous comments can be fixed by doing one of two things, changing the column to add Resident, therefore, fulling the requirements needed for the account.php to work properly or comment out line 64 in the account.php and not have it displayed bypassing it completely.

That first account creation and any created after that using the robust console will have this issue. This can present a problem with someone that has always created accounts in this fashion (due to the lack of sufficient scripts like this one or not enough knowledge in PHP and MySQL to create their own)

Now if there is a configuration part I missed somewhere in the OpenSim.ini, Robust,ini, GridCommon.ini or any other file, please let me know.

**Comment by emperorstarfinder** [comment]
Are you using the latest osregister from master?  

Our testing finds that the disconnect message and the UserTitle column in the DB being empty are not related.  the Disconnect message may more likely be related to a session() function issue and the cookies stored by your browser.  Clearing your browser's cache and cleaning up cookies the browser stores may help in this case.  

When logging in, the login.php script is not looking for UserTitle but rather FirstName, LastName, Password which is what it should be looking for.  The UserTitle column is just to allow users to know who is grid staff and who is a normal user without the need to show the GodLevel (i.e. UserLevel) in OpenSim 0.9.x and would have nothing at all to do with the website's ability to log in.

**Comment by Ramius1701** [comment]
I am using the latest master,

So the session is already commented out, again I made no changes to the config file with the exception of adding the DB info.

I do see where you're coming from on the login.php, although, the issue is after you log in and click on either "Manage your Account" button on the main screen or the drop-down menu "My account". Both of these lead to http://server.name/osregister/?page=account which is the account.php file. That's when it logs you out with the message "You are disconnected successfully Test User ..." This stems from the UserTitle column being empty for Test User as indicated by line 64 "empty($_SESSION['usertitle']) ||" once you comment out that line it will work for Test User. 

This does not happen with Tester User which was created when registering a new account "registration.php" and no changes made to line 64 as I have indicated that I had to do to get Test User to view the account page and that field is blank as it is in the DB. Tester User doesn't have this problem since this is filled in by the registration form.

Again Test User was created via the Robust.exe console when setting up the grid as needed to continue for the region to be set up in OpenSim.exe. Tester User was created via the registration form.

In summary, any grid setup that was creating accounts for its users via Robust.exe console may run into this problem with a lot of accounts and then try to use osregister. They will have to do one of two things currently either comment out line 64 in the accounts.php and not use it or modify the DB manually adding Resident, Owner, etc into the column for UserTitle for each user that doesn't have this entered for account.php to work as intended.

What I used to troubleshoot this is Windows 10, Ubuntu 20.04 LTS, Apache 2.4.1, PHP 7.4.4, MariaDB 10.4.12, Microsoft Edge 84.0.522.44, Chrome 84.0.4147.89, Firefox 78.0.2, Opera 69.0.3686.77, Notepad++ 7.8.9, OpenSim Version 0.9.1.1 Release and 0.9.2.0 Dev

**Comment by emperorstarfinder** [comment]
@Ramius1701 Try going into the UserAccounts DB table and under the user add `Resident` in the UserTitle column for that user and then try logging them in via the website.  IF that works then there would be an issue with that column in the PHP code.  The only problem is testing locally using the info you provided, I cannot reproduce that.  Additionally, it is working properly on my team's main grid and we have created users including our debugging account known as Test Debug via the console without a UserTitle and it has worked.  That will at least tell us, however, if the code is coughing on your server at that point though I really don't see why it should be.

**Comment by djphil** [comment]
@Ramius1701 : "Normally", in osregqister master there should be nothing to modify any file and the disconnection should be fixed because there is no more checking of the UserTitle.

@emperorstarfinder : It is planned in a future version of osregister to have a small version checking system. The frontend display can easily be modified by the end user and if it remains in the footer this is of course appreciated.

 I corrected the names of the tables. There shouldn't be a problem with that now.
Thank you for pointing the Opensim doc.

**Comment by emperorstarfinder** [comment]
@djphil Good stuff.  I updated the osregister on my team's internal testbed grid.  It seems you forgot to point the user accounts db table to whats in the config.php file in includes/login.php line 60.  I wound up having to edit that line to get logins to work on the testbed.  

As to the version number item, I will fork the repository and put the changes I am talking about in a branch to give you a better idea of what I mean.  I will update shortly.

**Comment by djphil** [comment]
Ho yes you're right, I completely skipped the login page. 
There it is, it is up to date now (5fe5ae43b2bca52c0761a4cebee98f9b30861395).
Thank you for pointing this out.

**Comment by emperorstarfinder** [comment]
@djphil to see what I was referring to about the version number item see: https://github.com/emperorstarfinder/osregister/commit/b91c9c08a5ee6b20a13907c7698bbf18d005e58d  I haven't tested it yet to make sure I didn't accidentally create a bug but it will demonstrate what I meant.

**Comment by djphil** [comment]
Thanks for your example.
I don't plan to make this change.
